server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1911,6 +1930,60 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
             XContentParserUtils.ensureExpectedToken(XContentParser.Token.END_OBJECT, parser.nextToken(), parser);
             return builder.build();
         }
+
+        private IndexMetadata reuseMappings(IndexMetadata indexMetadata) {
+            if (indexMetadata.mapping() == null) {
+                return indexMetadata;
+            }
+
+            String digest = indexMetadata.mapping().getSha256();
+            MappingMetadata entry = cache.get(digest);
+            if (entry != null) {
+                IndexMetadata.Builder imBuilder = new IndexMetadata.Builder(indexMetadata);
+                imBuilder.putMapping(entry);
+                return imBuilder.build();
+            } else {
+                cache.put(digest, indexMetadata.mapping());
+                return indexMetadata;
+            }
+        }
+
+        private void reuseMappings(IndexMetadata.Builder indexMetadataBuilder) {
+            if (indexMetadataBuilder.mapping() == null) {
+                return;
+            }
+
+            String digest = indexMetadataBuilder.mapping().getSha256();
+            MappingMetadata entry = cache.get(digest);
+            if (entry != null) {
+                indexMetadataBuilder.putMapping(entry);
+            } else {
+                cache.put(digest, indexMetadataBuilder.mapping());
+            }
+        }
+
+        private void purgeMappingCache(ImmutableOpenMap<String, IndexMetadata> indices) {
+            final var iterator = cache.entrySet().iterator();
+            while (iterator.hasNext()) {
```

Comment:
This seems needlessly hard :) Can't we just collect all the hash strings from the metadata into a set in one loop and then all keys from `cache` that aren't in that set? That way we're at least O(N)? :)

Reply:
Something like this: 3d2bf906b839b22330b047cc0441e4b3dfb0e040 ?

Reply:
++ perfect